### PR TITLE
[Turbopack] fix root and project path usages in a monorepo

### PR DESCRIPTION
**Pull request by sokra**
### What?

Improved trace source maps to return relative paths and always relative to process.cwd.

`turbopack://` urls in SourceMaps are always relative to the project root.

`file://` urls do no longer contain duplicate path segments.

Add a test case

Also:
* improve error when sourcemap lookup throws

Closes PACK-3671

### CODE REVIEW

**Comment by sokra** [comment]
* **#73553** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73553?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#72612** <a href="https://app.graphite.dev/github/pr/vercel/next.js/72612?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#73552** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73552?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/vercel/next.js/73552?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `canary`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: f70e8a5dcec4871c881529263dcf7c9667a020ea

`pnpm test-dev-turbo test/development/app-dir/owner-stack-react-missing-key-prop/owner-stack-react-missing-key-prop.test.ts` (turbopack)
- app-dir - owner-stack-react-missing-key-prop > should catch invalid element from on rsc component
- app-dir - owner-stack-react-missing-key-prop > should catch invalid element from on ssr client component

<details>
<summary>Expand output</summary>

  ● app-dir - owner-stack-react-missing-key-prop › should catch invalid element from on rsc component

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `app-dir - owner-stack-react-missing-key-prop should catch invalid element from on rsc component 1`

    Snapshot: "at Page (app/rsc/page.tsx (6:13))"
    Received: "at Page (./app/rsc/page.tsx (6:13))"

      26 |
      27 |       if (process.env.TURBOPACK) {
    > 28 |         expect(stackFramesContent).toMatchInlineSnapshot(
         |                                    ^
      29 |           `"at Page (app/rsc/page.tsx (6:13))"`
      30 |         )
      31 |         expect(source).toMatchInlineSnapshot(`

      at Object.toMatchInlineSnapshot (development/app-dir/owner-stack-react-missing-key-prop/owner-stack-react-missing-key-prop.test.ts:28:36)

  ● app-dir - owner-stack-react-missing-key-prop › should catch invalid element from on rsc component

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `app-dir - owner-stack-react-missing-key-prop should catch invalid element from on rsc component 2`

    - Snapshot  - 1
    + Received  + 1

    @@ -1,6 +1,6 @@
    - app/rsc/page.tsx (7:10) @ <anonymous>
    + ./app/rsc/page.tsx (7:10) @ <anonymous>

         5 |     <div>
         6 |       {list.map((item, index) => (
      >  7 |         <span>{item}</span>
           |          ^

      29 |           `"at Page (app/rsc/page.tsx (6:13))"`
      30 |         )
    > 31 |         expect(source).toMatchInlineSnapshot(`
         |                        ^
      32 |         "app/rsc/page.tsx (7:10) @ <anonymous>
      33 |
      34 |            5 |     <div>

      at Object.toMatchInlineSnapshot (development/app-dir/owner-stack-react-missing-key-prop/owner-stack-react-missing-key-prop.test.ts:31:24)

  ● app-dir - owner-stack-react-missing-key-prop › should catch invalid element from on ssr client component

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `app-dir - owner-stack-react-missing-key-prop should catch invalid element from on ssr client component 1`

    Snapshot: "at Page (app/ssr/page.tsx (8:13))"
    Received: "at Page (./app/ssr/page.tsx (8:13))"

      67 |       const source = await getRedboxSource(browser)
      68 |       if (process.env.TURBOPACK) {
    > 69 |         expect(stackFramesContent).toMatchInlineSnapshot(
         |                                    ^
      70 |           `"at Page (app/ssr/page.tsx (8:13))"`
      71 |         )
      72 |         expect(source).toMatchInlineSnapshot(`

      at Object.toMatchInlineSnapshot (development/app-dir/owner-stack-react-missing-key-prop/owner-stack-react-missing-key-prop.test.ts:69:36)

  ● app-dir - owner-stack-react-missing-key-prop › should catch invalid element from on ssr client component

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `app-dir - owner-stack-react-missing-key-prop should catch invalid element from on ssr client component 2`

    - Snapshot  - 1
    + Received  + 1

    @@ -1,6 +1,6 @@
    - app/ssr/page.tsx (9:9) @ <unknown>
    + ./app/ssr/page.tsx (9:9) @ <unknown>

         7 |     <div>
         8 |       {list.map((item, index) => (
      >  9 |         <p>{item}</p>
           |         ^

      70 |           `"at Page (app/ssr/page.tsx (8:13))"`
      71 |         )
    > 72 |         expect(source).toMatchInlineSnapshot(`
         |                        ^
      73 |         "app/ssr/page.tsx (9:9) @ <unknown>
      74 |
      75 |            7 |     <div>

      at Object.toMatchInlineSnapshot (development/app-dir/owner-stack-react-missing-key-prop/owner-stack-react-missing-key-prop.test.ts:72:24)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/integration/server-side-dev-errors/test/index.test.js
Comment (sokra) [inline test/integration/server-side-dev-errors/test/index.test.js:1]: @eps1lon see here

File: test/integration/server-side-dev-errors/test/index.test.js [hunk 1]
```diff
@@ -174,14 +174,14 @@ describe('server-side dev errors', () => {
...
         expect(stderrOutput).toStartWith(
-          ' ⨯ ../../test/integration/server-side-dev-errors/test/integration/server-side-dev-errors/pages/blog/[slug].js (6:3) @ getServerSideProps' +
-            '\n ⨯ ../../test/integration/server-side-dev-errors/test/integration/server-side-dev-errors/pages/blog/[slug].js (6:3) @ getServerSideProps'
+          ' ⨯ ../../test/integration/server-side-dev-errors/pages/blog/[slug].js (6:3) @ getServerSideProps' +
+            '\n ⨯ ../../test/integration/server-side-dev-errors/pages/blog/[slug].js (6:3) @ getServerSideProps'
         )
         expect(stderrOutput).toContain(
           ' ⨯ ReferenceError: missingVar is not defined' +
-            '\n    at getServerSideProps (./test/integration/server-side-dev-errors/pages/blog/[slug].js:6:3)' +
+            // TODO(sokra) weird ./ in front of the ../../
```
Comment (eps1lon) [inline test/integration/server-side-dev-errors/test/index.test.js:183]: Will probably be fixed in https://github.com/vercel/next.js/pull/71909

File: test/e2e/app-dir/non-root-project-monorepo/non-root-project-monorepo.test.ts [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+          .replaceAll(nonIsolatedPath, 'file://<full-path>/')
+          .replaceAll(isolatedPath, 'file://<full-path>/')
+      }
+
+      it('should work on RSC', async () => {
+        const browser = await next.browser('/source-maps-rsc')
+        await assertHasRedbox(browser)
+
+        if (isTurbopack) {
+          // TODO the function name should be hidden
```
Comment (eps1lon) [inline test/e2e/app-dir/non-root-project-monorepo/non-root-project-monorepo.test.ts:67]: Why?

**Review by eps1lon — APPROVED** [review]

**Comment by sokra** [comment]
### Merge activity

* **Dec 19, 6:56 AM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/73552).

**Pull request merged**